### PR TITLE
Optimize attestation processing

### PR DESCRIPTION
**Pull request by paulhauner**
## Issue Addressed

NA

## Proposed Changes

Provides a 95% reduction in the time it takes to process an attestation for best-case scenario (all nodes follow the same head and are synchronous within a slot). Also tolerates some degree of forking.

Gains are obtained via caching committees to avoid beacon state reads.

## Additional Info

- ~~Blocked on #817~~

### CODE REVIEW

File: beacon_node/beacon_chain/src/beacon_chain.rs [hunk 1]
```diff
@@ -776,214 +807,242 @@ impl<T: BeaconChainTypes> BeaconChain<T> {
...
-                    self.log,
-                    "Dropped attestation for unknown block";
-                    "block" => format!("{}", attestation.data.beacon_block_root)
-                );
+            return Ok(AttestationProcessingOutcome::UnknownHeadBlock {
+                beacon_block_root: attestation.data.beacon_block_root,
+            });
+        };
+
+        // TODO: check FFG stuff?
```
Comment (paulhauner) [inline beacon_node/beacon_chain/src/beacon_chain.rs]: TODO

File: eth2/state_processing/src/per_block_processing/signature_sets.rs [hunk 2]
```diff
@@ -170,6 +170,33 @@ pub fn indexed_attestation_signature_set<'a, 'b, T: EthSpec>(
     Ok(SignatureSet::new(signature, vec![signed_message]))
 }
 
+// TODO: try and combine this and the above function into one.
```
Comment (paulhauner) [inline eth2/state_processing/src/per_block_processing/signature_sets.rs]: TODO

File: beacon_node/network/src/service/tests.rs [hunk 3]
```diff
@@ -79,19 +51,7 @@ mod tests {
...
-                TestingSlotClock,
-                beacon_chain::eth1_chain::CachingEth1Backend<
-                    types::eth_spec::MinimalEthSpec,
-                    SimpleDiskStore<types::eth_spec::MinimalEthSpec>,
-                >,
-                types::eth_spec::MinimalEthSpec,
-                beacon_chain::events::NullEventHandler<types::eth_spec::MinimalEthSpec>,
-            >,
-        >(store);
+        let persisted_enrs = load_dht::<MemoryStore<MinimalEthSpec>, MinimalEthSpec>(store);
```
[the remark this answers is not in the document]
Comment (paulhauner) [reply beacon_node/network/src/service/tests.rs]: This stopped compiling due to some changes to the `BeaconChainBuilder`. Whilst fixing those I found it was overly verbose so I simplified it :)

[the remark this answers is not in the document]
Comment (michaelsproul) [reply beacon_node/network/src/service/tests.rs]: Nice!

File: beacon_node/beacon_chain/src/beacon_chain.rs [hunk 4]
```diff
@@ -756,223 +788,261 @@ impl<T: BeaconChainTypes> BeaconChain<T> {
             }
         }
 
+        metrics::stop_timer(timer);
         outcome
     }
 
     pub fn process_attestation_internal(
```
[the remark this answers is not in the document]
Comment (paulhauner) [reply beacon_node/beacon_chain/src/beacon_chain.rs]: I've been chatting to @AgeManning about this from time to time. My read is that we can merge this in now then come back to it once Age is ready.

**Comment by paulhauner** [comment]
Addressed all comments :)

File: beacon_node/beacon_chain/src/validator_pubkey_cache.rs [hunk 5]
```diff
@@ -0,0 +1,323 @@
...
+    }
+}
+
+impl ValidatorPubkeyCacheFile {
+    /// Creates a file for reading and writing.
+    pub fn create<P: AsRef<Path>>(path: P) -> Result<Self, Error> {
+        if path.as_ref().exists() {
+            Err(Error::FileExists)
+        } else {
+            File::create(path).map(Self).map_err(Error::IoError)
```
[the remark this answers is not in the document]
Comment (michaelsproul) [reply beacon_node/beacon_chain/src/validator_pubkey_cache.rs]: Can even delete the `if path.as_ref().exists()`

File: beacon_node/beacon_chain/src/beacon_chain.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (michaelsproul) [reply beacon_node/beacon_chain/src/beacon_chain.rs]: Sounds good

File: beacon_node/beacon_chain/src/validator_pubkey_cache.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (paulhauner) [reply beacon_node/beacon_chain/src/validator_pubkey_cache.rs]: Done, sorry!

[the remark this answers is not in the document]
Comment (michaelsproul) [reply beacon_node/beacon_chain/src/validator_pubkey_cache.rs]: Bellissimo

**Pull request merged**